clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -4298,6 +4314,18 @@ void handleFailure(Throwable throwable) {
                         }
                     }
                 }
+
+                @Override
+                boolean handleUnsupportedVersionException(UnsupportedVersionException exception) {
+                    if (supportsMaxTimestamp) {
+                        supportsMaxTimestamp = false;
+                        // check if there are any non MAX_TIMESTAMPS partitions left to be downgraded
+                        return partitionsToQuery.stream().anyMatch(
+                            t -> t.partitions().stream().anyMatch(
+                                p -> p.timestamp() != ListOffsetsRequest.MAX_TIMESTAMP));
+                    }
```

Comment:
Would it make sense to directly fail the future of the "max timestamp" requests here and to directly `partitionsToQuery` to only contains the remaining types? That would consolidate all the fall back logic here which is simpler to follow. Then, we could add another boolean `requireMaxTimestamp` to `ListOffsetsRequest.Builder.forConsumer` and we could directly pass `supportsMaxTimestamp` to it. This way, `createRequest` and `handleResponse` would remain unchanged. Did you already consider this?

Reply:
that makes sense, I've changed it to work that way now.